op-challenger/game/fault/trace/outputs/provider.go

```diff
@@ -0,0 +1,81 @@
+package outputs
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ethereum-optimism/optimism/op-challenger/game/fault/types"
+	"github.com/ethereum-optimism/optimism/op-service/client"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+var (
+	GetStepDataErr      = fmt.Errorf("GetStepData not supported")
+	AbsolutePreStateErr = fmt.Errorf("AbsolutePreState not supported")
+	PreStateRequestErr  = fmt.Errorf("Requested trace index is before prestate block")
+)
+
+type OutputRollupClient interface {
+	OutputAtBlock(ctx context.Context, blockNum uint64) (*eth.OutputResponse, error)
+}
+
+// OutputTraceProvider is a [types.TraceProvider] implementation that uses
+// output roots for given L2 Blocks as a trace.
+type OutputTraceProvider struct {
+	logger        log.Logger
+	rollupClient  OutputRollupClient
+	prestateBlock uint64
+}
+
+func NewTraceProvider(ctx context.Context, logger log.Logger, rollupRpc string, prestateBlock uint64) (*OutputTraceProvider, error) {
+	rollupClient, err := client.DialRollupClientWithTimeout(client.DefaultDialTimeout, logger, rollupRpc)
+	if err != nil {
+		return nil, err
+	}
+	return NewTraceProviderFromInputs(logger, rollupClient, prestateBlock), nil
+}
+
+func NewTraceProviderFromInputs(logger log.Logger, rollupClient OutputRollupClient, prestateBlock uint64) *OutputTraceProvider {
+	return &OutputTraceProvider{
+		logger:        logger,
+		rollupClient:  rollupClient,
+		prestateBlock: prestateBlock,
+	}
+}
+
+func (o *OutputTraceProvider) Get(ctx context.Context, traceIndex uint64) (common.Hash, error) {
```

Comment:
I think we should be shifting `traceIndex` by the `prestateBlock`. `traceIndex` 0 corresponds to `prestateBlock` and we also need to track the `postState` block because when `traceIndex > postStateBlockNumber - preStateBlockNumber` we need to return `postStateBlock`, not `traceIndex + preStateBlockNumber`'s block

Reply:
Good point, thanks!